xdp-dump/xdpdump.c

```diff
@@ -99,17 +113,115 @@ static struct prog_option xdpdump_options[] = {
-	uint64_t                   captured_packets;
-	uint64_t                   epoch_delta;
-	struct dumpopt            *cfg;
-	pcap_t                    *pcap;
-	pcap_dumper_t             *pcap_dumper;
+	uint64_t               missed_events;
+	uint64_t               last_missed_events;
+	uint64_t               captured_packets;
+	uint64_t               epoch_delta;
+	uint64_t               packet_id;
+	uint64_t               cpu_packet_id[MAX_CPUS];
+	struct dumpopt        *cfg;
+	pcap_t                *pcap;
+	pcap_dumper_t         *pcap_dumper;
+	struct xpcapng_dumper *pcapng_dumper;
 };
 
 bool    exit_xdpdump;
 pcap_t *exit_pcap;
 
+/*****************************************************************************
+ * get_if_speed()
+ *****************************************************************************/
+static uint64_t get_if_speed(struct iface *iface)
+{
+#define MAX_MODE_MASKS 10
+
+	int                                  fd;
+	struct ifreq                         ifr;
+	struct {
+		struct ethtool_link_settings req;
+		uint32_t                     modes[3 * MAX_MODE_MASKS];
+	} ereq;
+
+	if (iface == NULL)
+		return 0;
+
+	/* Open socket, and initialize structures. */
+	fd = socket(AF_INET, SOCK_DGRAM, 0);
+	if (fd < 0)
+		return 0;
+
+	memset(&ereq, 0, sizeof(ereq));
+	ereq.req.cmd = ETHTOOL_GLINKSETTINGS;
+
+	memset(&ifr, 0, sizeof(ifr));
+	strncpy(ifr.ifr_name, iface->ifname, sizeof(ifr.ifr_name) - 1);
+	ifr.ifr_data = (void *)&ereq;
+
+	/* First query the kernel to see how many masks we need to ask for. */
+	if (ioctl(fd, SIOCETHTOOL, &ifr) != 0)
+		goto error_exit;
+
+	if (ereq.req.link_mode_masks_nwords >= 0 ||
+	    ereq.req.link_mode_masks_nwords < -MAX_MODE_MASKS ||
+	    ereq.req.cmd != ETHTOOL_GLINKSETTINGS)
+		goto error_exit;
+
+	/* Now ask for the data set, and extract the speed in bps. */
+	ereq.req.link_mode_masks_nwords = -ereq.req.link_mode_masks_nwords;
+	if (ioctl(fd, SIOCETHTOOL, &ifr) != 0)
+		goto error_exit;
+
+	/* If speed is unknown return 0. */
+	if (ereq.req.speed == -1)
+		ereq.req.speed = 0;
+
+	close(fd);
+	return ereq.req.speed * 1000000;
+
+error_exit:
+	close(fd);
+	return 0;
+}
+
+/*****************************************************************************
+ * get_if_drv_info()
+ *****************************************************************************/
+static char *get_if_drv_info(struct iface *iface, char *buffer, size_t len)
+{
+	int                     fd;
+	char                   *r_buffer = NULL;
+	struct ifreq            ifr;
+	struct ethtool_drvinfo  info;
+
+	if (iface == NULL || buffer == NULL || len == 0)
+		return NULL;
+
+	fd = socket(AF_INET, SOCK_DGRAM, 0);
+	if (fd < 0)
+		return NULL;
+
+	memset(&info, 0, sizeof(info));
+	info.cmd = ETHTOOL_GDRVINFO;
+
+	memset(&ifr, 0, sizeof(ifr));
+	strncpy(ifr.ifr_name, iface->ifname, sizeof(ifr.ifr_name) - 1);
+	ifr.ifr_data = (void *)&info;
+
+	if (ioctl(fd, SIOCETHTOOL, &ifr) != 0)
+		goto exit;
+
+	snprintf(buffer, len,
```

Comment:
try_snprintf() + check return?

(This goes for all uses of snprintf)

Reply:
Will fix in next commit